### DISCUSSION TITLE
Can AFL++ used in QNX neutrino embedded system? If can how to?

### DISCUSSION THREAD

**Discussion by DarrenDaRui** [AFLplusplus/AFLplusplus#1674] [Q&A]
Hi maintainer,

Here I stuck with the problem to fuzz code in QNX neutrino embedded system. The reason is QNX use qcc as the compiler and not support LLVM related toolset.
Previously I use libFuzzer to fuzz both X86_64 and Arm platform in Linux, but failed to fuzz QNX as no LLVM toolset supported.
Could you help guide if AFL++ can do this or not? If so could guide me the main orientation?

Thanks very much.

**Comment by vanhauser-thc** [comment]
You are lucky, I looked into that last week.
It is very ficklish to do, and I found one roadblock I could not solve (but did not look into it very deeply).

Because you need to cross compile you have to compile afl++ for your host (easy), and then specifically the afl tools also for that target. BUT - you must configure both to use MMAP.

first for your host:
`sed -i 's/-lrt//' src/afl-cc.c`
`make TEST_MMAP=1 afl-as`
`make TEST_MMAP=1 -f GNUmakefile.llvm afl-cc`

then for your target:
`rm -f instrumentation/*.o`
`sed -i 's/-ldl//g' GNUmakefile`
`sed -i 's/-lrt//g' GNUmakefile`
`make AFL_NO_X86=1 TEST_MMAP=1 CC=x86_64-pc-nto-qnx7.1.0-gcc afl-fuzz afl-showmap afl-tmin`
then you copy afl-fuzz afl-showmap, afl-tmin and afl-cmin to your QNX.

Now to compile a target you need to do the following - the compiler to set in your source environment is `afl-gcc`, and additionally you have to set `AFL_CC=x86_64-pc-nto-qnx7.1.0-gcc AFL_CXX=x86_64-pc-nto-qnx7.1.0-g++` you also have to set `AFL_PATH` and point it to the AFL++ compilation directory.

Now the issue that I ran into is that the forkserver is not triggering and I do not know why. This could be worked around by setting `AFL_NO_FORKSRV=1` when starting afl-fuzz.

There might be further issues but I did not went deeper into this.

If you down further steps, please document them here!
:-)

**Reply by DarrenDaRui** [reply]
Hi Vanhauser, I hit build error in second part:
src/afl-fuzz-init.c:2927:17: error: 'SA_RESTART' undeclared (first use in this function); did you mean 'ERESTART'?
I check QNX source, the 'SA_RESTART' is unsupported.

#define SA_NODEFER      0x0010  /* Don't apply the sigmask when the signal is being delivered */
#define SA_NOCLDWAIT    0x0020  /* Do not generate zombies on unwaited child */
/* #define SA_RESTART      0x0040 (not supported yet) */    /* Restart the kernel call on signal return */
#if defined(__EXT_QNX)

Do you hit that? Some workaround be here?

**Reply by vanhauser-thc** [reply]
sure, use the dev branch

**Reply by DarrenDaRui** [reply]
Hi @vanhauser-thc 
I hope get your point well, but stills got some error when build target with afl-cc.

`ace/Fuzzing/AFL++/test/AFLplusplus/afl-compiler-rt.o
[D] DEBUG: rt=/local/mnt1/workspace/Fuzzing/AFL++/test/AFLplusplus/afl-compiler-rt.o obj_path=/local/mnt1/workspace/Fuzzing/AFL++/test/AFLplusplus
[D] DEBUG: cd '/local/mnt1/workspace/Fuzzing/AFL++/test'; 'aarch64-unknown-nto-qnx7.1.0-gcc' '-B' '/local/mnt1/workspace/Fuzzing/AFL++/test/AFLplusplus' 'helloworld.c' '-o' 'helloworld' '-g' '-O3' '-funroll-loops' '' '-D__AFL_HAVE_MANUAL_CONTROL=1' '-D__AFL_COMPILER=1' '-DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION=1' '-D__AFL_FUZZ_INIT()=int __afl_sharedmem_fuzzing = 1;extern unsigned int *__afl_fuzz_len;extern unsigned char *__afl_fuzz_ptr;unsigned char __afl_fuzz_alt[1048576];unsigned char *__afl_fuzz_alt_ptr = __afl_fuzz_alt;' '-D__AFL_COVERAGE()=int __afl_selective_coverage = 1;void __afl_coverage_discard();void __afl_coverage_skip();void __afl_coverage_on();void __afl_coverage_off();' '-D__AFL_COVERAGE_START_OFF()=int __afl_selective_coverage_start_off = 1;' '-D__AFL_COVERAGE_ON()=__afl_coverage_on()' '-D__AFL_COVERAGE_OFF()=__afl_coverage_off()' '-D__AFL_COVERAGE_DISCARD()=__afl_coverage_discard()' '-D__AFL_COVERAGE_SKIP()=__afl_coverage_skip()' '-D__AFL_FUZZ_TESTCASE_BUF=(__afl_fuzz_ptr ? __afl_fuzz_ptr : __afl_fuzz_alt_ptr)' '-D__AFL_FUZZ_TESTCASE_LEN=(__afl_fuzz_ptr ? *__afl_fuzz_len : (*__afl_fuzz_len = read(0, __afl_fuzz_alt_ptr, 1048576)) == 0xffffffff ? 0 : *__afl_fuzz_len)' '-D__AFL_LOOP(_A)=({ static volatile char *_B __attribute__((used,unused));  _B = (char*)"##SIG_AFL_PERSISTENT##"; extern int __afl_connected;__attribute__((visibility("default"))) int _L(unsigned int) __asm__("__afl_persistent_loop"); _L(__afl_connected ? _A : 1); })' '-D__AFL_INIT()=do { static volatile char *_A __attribute__((used,unused));  _A = (char*)"##SIG_AFL_DEFER_FORKSRV##"; __attribute__((visibility("default"))) void _I(void) __asm__("__afl_manual_init"); _I(); } while (0)' ''
aarch64-unknown-nto-qnx7.1.0-gcc: error: : No such file or directory
aarch64-unknown-nto-qnx7.1.0-gcc: error: : No such file or directory
`
Then I check the source of afl-cc, seems execvp(cc_params[0], (char **)cc_params); failed to work with QNX cross-compiler aarch64-unknown-nto-qnx7.1.0-gcc. Don't you ever encounter this error?
Another question, afl-cc should be x86_64 format to cross-compile target with AFL_CC and AFL_CXX set by QNX compiler, is that correct understanding? 
Looking for you answer. Thank you!

**Reply by vanhauser-thc** [reply]
is aarch64-unknown-nto-qnx7.1.0-gcc the compile that you have available? maybe you have 7.0.0. or you do not have it in your PATH.
afl-cc for your host linux system. it calls the cross compiler from qnx (via AFL_CC/AFL_CXX) with the right options to perform the instrumentation.

**Reply by DarrenDaRui** [reply]
Hi @vanhauser-thc , thanks for your reply. Finally I maybe found the reason, which replace qcc with aarch64-unknown-nto-qnx7.1.0-gcc. I don't know how it could be failed to find aarch64-unknown-nto-qnx7.1.0-gcc compiler, however, it's just there in $PATH also. Anyway it can be workaround with qcc. Then I encounter failure "cc1plus: error: command line option '-B /local/mnt1/workspace/Fuzzing/AFL++/AFLplusplus' is valid for the driver but not for C++", then I modify afl-cc code to remove -B and obj_path. Finally I got the target binary which can run on my QNX os board.

One more question, what's the best solution for AFL++ to do whitebox fuzzing as the source code is availabl. And mostly I do fuzzing for API test, which I invoke API in fuzzer harness with libFuzzer before. What can I do with AFL++ engine?

**Reply by vanhauser-thc** [reply]
you can do the same with afl++, same -fsanitize=fuzzer and you use afl-clang-fast - but it will not work with afl-gcc out of the box.
You can do your own easily if have LLVMFuzzerTestOneInput() harnesses for qnx. just do a main that reads from stdin and calls LLVMFuzzerTestOneInput.

if you remove the -B ... then it might compile but the target will not be instrumented for fuzzing.

**Reply by DarrenDaRui** [reply]
Hi @vanhauser-thc ,

I check what does -B means with clang/gcc, but can't find the clue how it will affect the code instrumented. Still can't figure out why aarch64-unknown-nto-qnx7.1.0-gcc will faile for afl-cc cross-compile target.
There still some points not clear for me to fuzz in QNX on-target. 
1. As you may know, QNX compiler doesn't support sanitizer like ASAN, so how do we monitor system under test once crash happened. Is there possibile debug with gdb to run the crash seed?
2. Thanks for your explain or how to use libFuzzer harnesses based on AFL++, but I would like to know what's the better way to fuzz API with AFL++, I still many sample which run binary with afl-fuzz directory as black-box fuzzing. I think AFL++ must have some better way for  white-box fuzzing, could you guide me?

**Reply by DarrenDaRui** [reply]
> is aarch64-unknown-nto-qnx7.1.0-gcc the compile that you have available? maybe you have 7.0.0. or you do not have it in your PATH. afl-cc for your host linux system. it calls the cross compiler from qnx (via AFL_CC/AFL_CXX) with the right options to perform the instrumentation.

@vanhauser-thc 
Find the rootcause, after sed -i 's/-lrt//' src/afl-cc.c, some "" added into aarch64-unknown-nto-qnx7.1.0-gcc compile cmdline, but it will caused some unknown error. After remove cc_params[cc_par_cnt++] = ""; something like this, the error is gone. But have no idea for why it happened.

**Reply by vanhauser-thc** [reply]
ah sorry. when I tried it I commented it out, for you I scripted it

**Reply by DarrenDaRui** [reply]
@vanhauser-thc 

I encounter No instrumentation detected issue when run afl-fuzz on QNX on-target. Is there any possibile I should link QNX 'as' tool aarch64-unknown-nto-qnx7.1.0-as to afl-as in the same way?

**Reply by vanhauser-thc** [reply]
that is where I got stuck too when I still had access to the system. you can try that yes

**Reply by DarrenDaRui** [reply]
@vanhauser-thc 

![image](https://user-images.githubusercontent.com/25190976/227448254-1d36734d-0c0a-4cb3-a8c2-a61763502c9d.png)

After link qnx aarch64-unknown-nto-qnx7.1.0-as to afl-as, afl-as can be invoked. However encounter error with ld stage. Do you have any idea?

**Reply by DarrenDaRui** [reply]
Hi @vanhauser-thc ,

Seems the assemble code in afl-as.h is x86 format, not suitable for aarch64 format. I got lot of "Error: unknown mnemonic `movq' -- " somethig like this. Really hard to go on, could you help guide me a little?

**Reply by vanhauser-thc** [reply]
oh damn yes. I didnt know that you are on aarch64. my QNX was on x86_x64 ...
you now have 3 choices:
1. rewrite the assembly for arm64
2. compile and fuzz on linux (best)
3. do system emulated fuzzing (e.g. nyx_mode or unicorn_mode)

**Reply by DarrenDaRui** [reply]
@vanhauser-thc 
Fine, thanks for your idea. 
Is that 2nd method means do "off-target" fuzzing based on Linux X86_64 platform?

**Reply by vanhauser-thc** [reply]
yes, you "just" compile it on linux ... mock whatever hardware would be there etc.

**Comment by vanhauser-thc** [comment]
Also there is this paper: https://ics-cert.kaspersky.com/media/Kaspersky-ICS-CERT-Dynamic-analysis-of-firmware-components-in-iot-devices-En.pdf
which shows how to use AFL++ to fuzz QNX binaries/libs with Unicorn mode

**Reply by DarrenDaRui** [reply]
Thanks very much for you reply, I will go through this paper. BTW, may I know it will work in QNX run-time env, I mean do on-target fuzzing.
Some QNX fuzzing may utilize QEMU platform to boot up QNX, but that's not we want.

**Reply by vanhauser-thc** [reply]
what I wrote in the other answer is for on-target fuzzing (obivously, if you look at it)

**Reply by DarrenDaRui** [reply]
Yes, just found the older one. I will learn your way and verify. Thanks again!
